[froniuswattpilot] Initial contribution

florian-h05 opened the pull request:
This adds a new binding that allows to integrate Fronius Wattpilot wallboxes into openHAB.

It has been tested with the following models:

- Fronius Wattpilot Home 11J (tested by @sihui62)
- Fronius Wattpilot Home 22J (tested myself)

lsiepel commented on bundles/org.openhab.binding.froniuswattpilot/src/main/java/org/openhab/binding/froniuswattpilot/internal/FroniusWattpilotHandler.java:
```diff
@@ -0,0 +1,308 @@
+    @Override
+    public void initialize() {
+        config = getConfigAs(FroniusWattpilotConfiguration.class);
+        updateStatus(ThingStatus.UNKNOWN);
```
Minor comment: move this to 173, just before you start the connection.

lsiepel commented on bundles/org.openhab.binding.froniuswattpilot/src/main/java/org/openhab/binding/froniuswattpilot/internal/FroniusWattpilotHandler.java:
```diff
@@ -0,0 +1,308 @@
+        }
+
+        try {
+            client.connect(config.hostname, config.password);
```
i guess this starts a http reqauest/response. That should be run from an async thread. The initialize method should be non-blocking and exit fast

florian-h05 replied:
This starts the Jetty WebSocketClient and calls the asynchronous connect method of it. It does not perform a network requests synchronously.
You can check the implementation at https://github.com/florian-h05/wattpilot4j/blob/89b4ac4c897e94728ae4790d7839bfe565d927fc/src/main/java/com/florianhotze/wattpilot/WattpilotClient.java#L150.

lsiepel commented on bundles/org.openhab.binding.froniuswattpilot/src/main/java/org/openhab/binding/froniuswattpilot/internal/FroniusWattpilotMDNSDiscoveryServiceParticipant.java:
```diff
@@ -0,0 +1,110 @@
+                String ip = addresses[0];
+                return DiscoveryResultBuilder.create(uid) //
+                        .withProperties(Map.of( //
+                                "hostname", ip)) //
+                        .withLabel(service.getName()) //
+                        .withRepresentationProperty("hostname") //
```
Woudl be better to extract the configuration key to a binding level string constant

lsiepel commented on bundles/org.openhab.binding.froniuswattpilot/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,209 @@
+	<thing-type id="wattpilot">
+		<label>Fronius Wattpilot</label>
+		<description>Fronius Wattpilot wallbox</description>
+
```
Could you add an equipment tag. 
For reference see: https://next.openhab.org/docs/developer/bindings/thing-xml.html#thing-semantic-equipment-tags

lsiepel reviewed (changes requested):
Thanks for this contribution!.
Nice binding to review, good quality and just a few comments:

florian-h05 commented:
@lsiepel I addressed your review. The build currently fails because we have to wait for Maven Central completing deploying the latest wattpilot4j version to its CDN.

jlaur commented on bundles/org.openhab.binding.froniuswattpilot/src/main/java/org/openhab/binding/froniuswattpilot/internal/FroniusWattpilotHandler.java:
```diff
@@ -0,0 +1,311 @@
+    public FroniusWattpilotHandler(Thing thing, HttpClient httpClient) {
+        super(thing);
+        client = new WattpilotClient(httpClient);
+        client.addListener(this);
```
I'm wondering if it's a problem that there's no `removeListener` available? Usually the handler would add itself as a listener in `initialize` and remove itself in `dispose`. I guess by subscribing outside of these lifecycle methods, we could receive events even when not initialized, causing the handler to update channels and Thing status?

florian-h05 replied:
Hmm, the question is what is the lifecycle of the handler here.
If the handler instance is removed after calling dispose, we don't need to remove the listteners as the client instance will also be removed together with the handler.
My wattpilot4j library and the binding itself follow the structure of the Harmony Hub binding, I am pretty sure there no listener is removed.

florian-h05 replied:
Internally the listeners are stored in a HashSet, so they cannot be added multiple times.
Once you call disconnect, onDisconnect will trigger once and then you are disconnected and won’t receive any updates.

jlaur commented on bundles/org.openhab.binding.froniuswattpilot/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,210 @@
+				<description>Energy charged in total</description>
+			</channel>
+			<channel id="l1-power" typeId="system.electric-power">
+				<label>Phase 1 Power</label>
```
You might want to override description as well, otherwise it will default to "Current electric power" (for all channels referring to `system.electric-power`).

jlaur commented on bundles/org.openhab.binding.froniuswattpilot/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,210 @@
+		<description>The mode of charging</description>
+		<category>BatteryLevel</category>
+		<tags>
+			<tag>Control</tag>
```
```suggestion
			<tag>Control</tag>
			<tag>Mode</tag>
```

jlaur commented on bundles/org.openhab.binding.froniuswattpilot/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,210 @@
+		<autoUpdatePolicy>veto</autoUpdatePolicy>
+	</channel-type>
+	<channel-type id="charging-current">
+		<item-type>Number:ElectricCurrent</item-type>
```
I don't remember system default unit for current, but in case it's not A:
```suggestion
		<item-type unitHint="A">Number:ElectricCurrent</item-type>
```

florian-h05 replied:
Ampere (A) is the default unit, just checked the docs :+1:

jlaur commented on bundles/org.openhab.binding.froniuswattpilot/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,210 @@
+		<item-type>Switch</item-type>
+		<label>Charging Allowed</label>
+		<description>Allow or forbid charging</description>
+		<category>BatteryLevel</category>
```
I think `BatteryLevel` is intended for battery level channels, i.e. to visually show the charge level based on the channel state. Since this is a `Switch` channel, you might want to remove this category, or perhaps use `Battery` instead. Same for others below.

jlaur commented on bundles/org.openhab.binding.froniuswattpilot/README.md:
```diff
@@ -0,0 +1,95 @@
+// Metrics phase 3
+Number:Power              Wattpilot_Garage_Phase_3_Power               "Phase 3 Power [%.2f kW]"               <Energy>        (Wattpilot_Garage)  ["Measurement", "Power"]    {channel="froniuswattpilot:wattpilot:garage:metrics#l3-power", unit="kW"}
+Number:ElectricPotential  Wattpilot_Garage_Phase_3_Voltage             "Phase 3 Voltage [%d V]"                <Energy>        (Wattpilot_Garage)  ["Measurement", "Voltage"]  {channel="froniuswattpilot:wattpilot:garage:metrics#l3-voltage", unit="V"}
+Number:ElectricCurrent    Wattpilot_Garage_Phase_3_Current             "Phase 3 Current [%.1f A]"              <Energy>        (Wattpilot_Garage)  ["Current", "Measurement"]  {channel="froniuswattpilot:wattpilot:garage:metrics#l3-current", unit="A"}
```
```suggestion
Number:ElectricCurrent    Wattpilot_Garage_Phase_3_Current             "Phase 3 Current [%.1f A]"              <Energy>        (Wattpilot_Garage)  ["Measurement", "Current"]  {channel="froniuswattpilot:wattpilot:garage:metrics#l3-current", unit="A"}
```

jlaur commented on bundles/org.openhab.binding.froniuswattpilot/README.md:
```diff
@@ -0,0 +1,95 @@
+// Metrics phase 2
+Number:Power              Wattpilot_Garage_Phase_2_Power               "Phase 2 Power [%.2f kW]"               <Energy>        (Wattpilot_Garage)  ["Measurement", "Power"]    {channel="froniuswattpilot:wattpilot:garage:metrics#l2-power", unit="kW"}
+Number:ElectricPotential  Wattpilot_Garage_Phase_2_Voltage             "Phase 2 Voltage [%d V]"                <Energy>        (Wattpilot_Garage)  ["Measurement", "Voltage"]  {channel="froniuswattpilot:wattpilot:garage:metrics#l2-voltage", unit="V"}
+Number:ElectricCurrent    Wattpilot_Garage_Phase_2_Current             "Phase 2 Current [%.1f A]"              <Energy>        (Wattpilot_Garage)  ["Current", "Measurement"]  {channel="froniuswattpilot:wattpilot:garage:metrics#l2-current", unit="A"}
```
```suggestion
Number:ElectricCurrent    Wattpilot_Garage_Phase_2_Current             "Phase 2 Current [%.1f A]"              <Energy>        (Wattpilot_Garage)  ["Measurement", "Current"]  {channel="froniuswattpilot:wattpilot:garage:metrics#l2-current", unit="A"}
```

jlaur commented on bundles/org.openhab.binding.froniuswattpilot/README.md:
```diff
@@ -0,0 +1,95 @@
+// Metrics phase 1
+Number:Power              Wattpilot_Garage_Phase_1_Power               "Phase 1 Power [%.2f kW]"               <Energy>        (Wattpilot_Garage)  ["Measurement", "Power"]    {channel="froniuswattpilot:wattpilot:garage:metrics#l1-power", unit="kW"}
+Number:ElectricPotential  Wattpilot_Garage_Phase_1_Voltage             "Phase 1 Voltage [%d V]"                <Energy>        (Wattpilot_Garage)  ["Measurement", "Voltage"]  {channel="froniuswattpilot:wattpilot:garage:metrics#l1-voltage", unit="V"}
+Number:ElectricCurrent    Wattpilot_Garage_Phase_1_Current             "Phase 1 Current [%.1f A]"              <Energy>        (Wattpilot_Garage)  ["Current", "Measurement"]  {channel="froniuswattpilot:wattpilot:garage:metrics#l1-current", unit="A"}
```
```suggestion
Number:ElectricCurrent    Wattpilot_Garage_Phase_1_Current             "Phase 1 Current [%.1f A]"              <Energy>        (Wattpilot_Garage)  ["Measurement", "Current"]  {channel="froniuswattpilot:wattpilot:garage:metrics#l1-current", unit="A"}
```

jlaur commented on bundles/org.openhab.binding.froniuswattpilot/README.md:
```diff
@@ -0,0 +1,95 @@
+// Metrics total
+Number:Power              Wattpilot_Garage_Total_Power                 "Total Power [%.2f kW]"                 <Energy>        (Wattpilot_Garage)  ["Measurement", "Power"]    {channel="froniuswattpilot:wattpilot:garage:metrics#power", unit="kW"}
+Number:Energy             Wattpilot_Garage_Charged_Energy              "Charged Energy [%.2f kWh]"             <Energy>        (Wattpilot_Garage)  ["Energy", "Measurement"]   {channel="froniuswattpilot:wattpilot:garage:metrics#energy-session", unit="kWh"}
+Number:Energy             Wattpilot_Garage_Total_Charged_Energy        "Total Charged Energy [%.0f kWh]"       <Energy>        (Wattpilot_Garage)  ["Energy", "Measurement"]   {channel="froniuswattpilot:wattpilot:garage:metrics#energy-total", unit="kWh"}
```
```suggestion
Number:Energy             Wattpilot_Garage_Total_Charged_Energy        "Total Charged Energy [%.0f kWh]"       <Energy>        (Wattpilot_Garage)  ["Measurement", "Energy"]   {channel="froniuswattpilot:wattpilot:garage:metrics#energy-total", unit="kWh"}
```

jlaur commented on bundles/org.openhab.binding.froniuswattpilot/README.md:
```diff
@@ -0,0 +1,95 @@
+// Control
+Switch                    Wattpilot_Garage_Charging_Allowed            "Charging Allowed"                      <BatteryLevel>  (Wattpilot_Garage)  ["Control"]                 {channel="froniuswattpilot:wattpilot:garage:control#charging-allowed"}
+String                    Wattpilot_Garage_Charging_Mode               "Charging Mode"                         <BatteryLevel>  (Wattpilot_Garage)  ["Control"]                 {channel="froniuswattpilot:wattpilot:garage:control#charging-mode"}
+Number:ElectricCurrent    Wattpilot_Garage_Charging_Current            "Charging Current [%d A]"               <Energy>        (Wattpilot_Garage)  ["Current", "Setpoint"]     {channel="froniuswattpilot:wattpilot:garage:control#charging-current", unit="A"}
+Number:Power              Wattpilot_Garage_PV_Surplus_Power_Threshold  "PV Surplus Power Threshold [%.1f kW]"  <SolarPlant>    (Wattpilot_Garage)  ["Power", "Setpoint"]       {channel="froniuswattpilot:wattpilot:garage:control#pv-surplus-threshold", unit="kW"}
```
```suggestion
Number:ElectricCurrent    Wattpilot_Garage_Charging_Current            "Charging Current [%d A]"               <Energy>        (Wattpilot_Garage)  ["Setpoint", "Current"]     {channel="froniuswattpilot:wattpilot:garage:control#charging-current", unit="A"}
Number:Power              Wattpilot_Garage_PV_Surplus_Power_Threshold  "PV Surplus Power Threshold [%.1f kW]"  <SolarPlant>    (Wattpilot_Garage)  ["Setpoint", "Power"]       {channel="froniuswattpilot:wattpilot:garage:control#pv-surplus-threshold", unit="kW"}
```

jlaur commented:
@florian-h05 - thanks for addressing my comments. 👍 Can you also have a look at the compiler warnings? @lsiepel - feel free to merge when it LGTY. 😉

florian-h05 commented:
@jlaur Compiler warnings are fixed now.

lsiepel commented:
Might be missing a git push? As i still see some build warnings:
```
[WARNING] D:\oh-dev\openhab-addons\bundles\org.openhab.binding.fronius\src\main\java\org\openhab\binding\fronius\internal\api\FroniusBatteryControl.java:[132,14] Potential null pointer access: The variable response may be null at this location
[WARNING] D:\oh-dev\openhab-addons\bundles\org.openhab.binding.fronius\src\main\java\org\openhab\binding\fronius\internal\api\FroniusBatteryControl.java:[210,64] Potential null pointer access: The method toUnit(Unit) may return null
[WARNING] D:\oh-dev\openhab-addons\bundles\org.openhab.binding.fronius\src\main\java\org\openhab\binding\fronius\internal\api\FroniusBatteryControl.java:[242,14] Potential null pointer access: The variable response may be null at this location
[WARNING] D:\oh-dev\openhab-addons\bundles\org.openhab.binding.fronius\src\main\java\org\openhab\binding\fronius\internal\api\FroniusUnauthorizedException.java:[20,14] The serializable class FroniusUnauthorizedException does not declare a static final serialVersionUID field of type long
[WARNING] D:\oh-dev\openhab-addons\bundles\org.openhab.binding.fronius\src\main\java\org\openhab\binding\fronius\internal\handler\FroniusSymoInverterHandler.java:[101,51] Potential null pointer access: The method getBridge() may return null
[WARNING] D:\oh-dev\openhab-addons\bundles\org.openhab.binding.fronius\src\main\java\org\openhab\binding\fronius\internal\handler\FroniusSymoInverterHandler.java:[114,51] Potential null pointer access: The method getBridge() may return null
[WARNING] D:\oh-dev\openhab-addons\bundles\org.openhab.binding.fronius\src\main\java\org\openhab\binding\fronius\internal\handler\FroniusSymoInverterHandler.java:[383,64] javax.measure.Unit is a raw type. References to generic type javax.measure.Unit<Q> should be parameterized
[WARNING] D:\oh-dev\openhab-addons\bundles\org.openhab.binding.fronius\src\main\java\org\openhab\binding\fronius\internal\handler\FroniusSymoInverterHandler.java:[385,25] Type safety: The constructor org.openhab.core.library.types.QuantityType(java.lang.Number, javax.measure.Unit) belongs to the raw type org.openhab.core.library.types.QuantityType. References to generic type org.openhab.core.library.types.QuantityType<T> should be parameterized
[WARNING] D:\oh-dev\openhab-addons\bundles\org.openhab.binding.fronius\src\main\java\org\openhab\binding\fronius\internal\handler\FroniusSymoInverterHandler.java:[385,25] Type safety: The expression of type org.openhab.core.library.types.QuantityType needs unchecked conversion to conform to org.openhab.core.library.types.QuantityType<capture#17-of ?>
```

Also get some SAT errors that need to be fixed.

florian-h05 commented:
You talk about the Fronius Binding, this is the Fronius Wattpilot Binding.

lsiepel commented:
> You talk about the Fronius Binding, this is the Fronius Wattpilot Binding.

My mistake when building it locally. Will look at the right one again 🫣

lsiepel commented:
You could add the binding's logo to the openHAB website. See https://www.openhab.org/docs/developer/addons/#add-your-add-on-s-logo-to-the-openhab-website
